tephra-core/src/main/java/co/cask/tephra/persist/HDFSTransactionLogReaderV2.java

```diff
@@ -0,0 +1,131 @@
+  private static final byte[] KEY_BYTES = TxConstants.TransactionLog.NUM_ENTRIES_APPENDED.getBytes();
+
+  private final Queue<TransactionEdit> transactionEdits;
+  private final DataOutputBuffer rawKey;
+  private final ByteArrayOutputStream value;
+  private final DataOutputStream outputStream;
+  private final LongWritable key;
+
+  public HDFSTransactionLogReaderV2(SequenceFile.Reader reader) {
+    super(reader);
+    this.transactionEdits = new ArrayDeque<>();
+    this.key = new LongWritable();
+    this.rawKey = new DataOutputBuffer();
+    this.value = new ByteArrayOutputStream(Integer.SIZE);
+    this.outputStream = new DataOutputStream(value);
+  }
+
+  @Override
+  public void close() throws IOException {
+    outputStream.close();
+    value.close();
+    rawKey.close();
+    super.close();
+  }
+
+  @Override
+  public TransactionEdit next() throws IOException {
+    return next(null);
+  }
+
+  @Override
+  public TransactionEdit next(TransactionEdit reuse) throws IOException {
+    if (isClosed()) {
+      return null;
+    }
+
+    if (!transactionEdits.isEmpty()) {
+      return transactionEdits.remove();
+    }
+
+    // Fetch the 'marker' and read 'marker' number of edits
+    populateTransactionEdits();
+    return transactionEdits.poll();
+  }
+
+  private void populateTransactionEdits() throws IOException {
+    // read the marker to determine numEntries to read.
+    int numEntries = fetchMarker();
+    for (int i = 0; i < numEntries; i++) {
+      TransactionEdit edit = new TransactionEdit();
+      try {
+        if (reader.next(key, edit)) {
+          transactionEdits.add(edit);
+        } else {
+          throw new EOFException("Attempt to read TransactionEdit failed.");
+        }
+      } catch (EOFException e) {
+        // we have reached EOF before reading back numEntries, we clear the partial list and return.
+        LOG.warn("Reached EOF in log before reading {} entries. Ignoring all {} edits since the last marker",
+                 numEntries, transactionEdits.size(), e);
+        transactionEdits.clear();
+      }
+    }
+  }
+
+  private int fetchMarker() throws IOException {
+    SequenceFile.ValueBytes valueBytes = reader.createValueBytes();
+    rawKey.reset();
+    int status = reader.nextRaw(rawKey, valueBytes);
+
+    //Check if the marker key is valid
+    if (status != -1 && isMarkerValid()) {
+      valueBytes.writeUncompressedBytes(outputStream);
+      outputStream.flush();
+      return Ints.fromByteArray(value.toByteArray());
+    }
+
+    // EOF not reached but couldn't read marker, then log an error
+    if (status != -1) {
+      LOG.error("Invalid key for num entries appended found {}, expected : {}",
+                rawKey.getData(), TxConstants.TransactionLog.NUM_ENTRIES_APPENDED);
+    }
+    return -1;
+  }
+
+  private boolean isMarkerValid() {
+    return rawKey.getLength() == KEY_BYTES.length && Bytes.indexOf(rawKey.getData(), KEY_BYTES) != -1;
```

Comment:
`rawKey.getData()` returns a `byte[]` larger than `KEY_BYTES.length`, so we need to make sure that the `KEY_BYTES` matches just the beginning of it, and not that it appears anywhere in the `rawKey.getData()`, right?